Google.Solutions.Compute/Auth/RegistryStore.cs

```diff
@@ -0,0 +1,139 @@
+﻿//
+// Copyright 2019 Google LLC
+//
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+// 
+//   http://www.apache.org/licenses/LICENSE-2.0
+// 
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+//
+
+using Google.Apis.Json;
+using Google.Apis.Util;
+using Google.Apis.Util.Store;
+using Microsoft.Win32;
+using System;
+using System.Security.Cryptography;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Google.Solutions.Compute.Auth
+{
+    /// <summary>
+    /// Authentication data store that encrypts data by using DPAPI and 
+    /// stores the cyphertext in the registry.
+    /// 
+    /// Tokens tend to be slightly too long to be stored in the Windows 
+    /// Credential Store (CredRead/CredWrite), so the registry is more 
+    /// flexible.
+    /// </summary>
+    public class RegistryStore : IDataStore, IDisposable
+    {
+        private static readonly Task CompletedTask = Task.FromResult(0);
+
+        private readonly RegistryHive hive;
+        private readonly RegistryKey key;
+
+        public RegistryStore(RegistryHive hive, string keyPath)
+        {
+            this.hive = hive;
+            this.key = RegistryKey.OpenBaseKey(hive, RegistryView.Default)
+                        .CreateSubKey(keyPath, true);
+        }
+
+        private DataProtectionScope ProtectionScope =>
+            this.hive == RegistryHive.LocalMachine
```

Comment:
cool;  so the credentials don't 'migrate'  stay on one machine.  nice; probably should mention that in the docs.

Reply:
I kept the API generic to support both, 
* machine-specific credentials (in HKLM, non-roaming, encrypted using a machine key)
* user-specific credentials (in HKCU, roaming, encrypted using a user key)

But for storing tokens, I'll have to use HKCU. That means I'll also have to use the (roaming) user key to encrypt them.